# Slope and Derivatives

$\frac{df}{dx}=\frac{f(x+\delta)-f(x)}{\delta}$

Our function for the following returns $x(x-1)$ and take $\delta=10^{-2}$


Derivative of $\frac{d}{dx}x(x-1) = \frac{d}{dx} (x^2-x)$ is $(2x-1)$, so this valiue at $x=1$ is 1 

In [48]:
# import all your libraries
import numpy as np
import cmath
import math

In [49]:
def f(x):
    return x*(x-1)

delta = 1e-10 # 10**(-2)
x = 1
df = (f(x+delta)-f(x))/delta

print(delta, df)

1e-10 1.000000082840371


## Exercise: The position of an object is given as $x(t)=2t^3-3t^2+4t-6$

Velocity, which is derivative of the position is $v(t)=6t^2-6t+4$ which at time (2 seconds) $v(2)=16 m/s$



In [50]:
def x(t):
    return 2*t**3-3*t**2+4*t-6

delta = 1e-14 # 10**(-2)
t = 2
dx = (x(t+delta)-x(t))/delta

print(delta, dx)

1e-14 16.342482922482304


For $\delta=1e-14$, the solution provides $16.342482922482304$ as an aswer that means it gets worse. This is so because the two numbers (on the numerator) are very close to each other we get fractional error.

# Questions 
What is the acceleration? $a(t)=12t-6$ and this at $t =2$ seconds gives $a(2)=18 m/s^2$

In [51]:
def v(t):
    return 6*t**2-6*t+4

delta = 1e-8 # 10**(-2)
t = 2
dv = (v(t+delta)-v(t))/delta

print(delta, dv)

1e-08 18.000000068241206


# Accuracy (int and float)

factorial $n!=n(n-1)(n-2)\ldots[n-(n-1)]$

In [52]:
fac = 4 #int
for i in range(1,fac):
    fac *= i #3!= 3*2*1
print(fac)

24


# Quadratic Equations $ax^2+bx+c$, solution is $\frac{-b\pm\sqrt{b^2-4ac}}{2a}$

Roots $r_1*r_2 = 1$

In [54]:

a = float(input("Enter a: "))
b = float(input("Enter b: "))
c = float(input("Enter c: "))

squareroot = np.sqrt(b*b-4*a*c)

if b>0.0: #
    r1=(-b-squareroot)/(2*a) #r1
    r2= 2*a/(-b-squareroot)
else:
    r1=(-b+squareroot)/(2*a) #r2
    r2= 2*a/(-b+squareroot)
    
print(r1,r2)

Enter a: 1
Enter b: 5
Enter c: 2
-4.561552812808831 -0.21922359359558485


# On Jasmine's question 

To produce cases where the discriminant produces different results including complex number

$\frac{-b\pm\sqrt{b^2-4ac}}{2a}$ 

Discriminant rules

* If $b^2-4ac < 0$ the quadratic equation has no real solution, its soilution is complex number
* If $b^2-4ac = 0 $ the quadratic equation has only one solution, in fact we have two real and same roots/solutions.
* $b^2-4ac > 0$ the quadratic equation has two real and different solution

In [55]:

a = float(input("Enter a: "))
b = float(input("Enter b: "))
c = float(input("Enter c: "))

squareroot = np.sqrt(b*b-4*a*c) 

if squareroot > 0.0: #
    print("Gives two real and distinct roots")
    r1=(-b-squareroot)/(2*a) #r1
    r2=(-b+squareroot)/(2*a)
    print(r1,r2)
    
elif squareroot == 0.0: #
    print("Gives one real root (or two same roots)")
    r1=(-b)/(2*a) #r1
    r2=(-b)/(2*a)
    print(r1,r2)
    
else:
    r1=(-b)/(2*a) #r2
    r1=(-b)/(2*a)
    print("Gives two Complex Roots")
    #print(- b / (2 * a), + i, squareroot)
    #print(- b / (2 * a), - i, squareroot)
    print(r1,r1)





Enter a: 2
Enter b: 4
Enter c: 6
Gives two Complex Roots
-1.0 -1.0


/Users/ramlijufar/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:5: RuntimeWarning: invalid value encountered in sqrt
  """


# But we can improve the code as follows

What if we want to show the complex number

In [56]:

a = float(input("Enter a: "))
b = float(input("Enter b: "))
c = float(input("Enter c: "))

squareroot = np.sqrt(b*b-4*a*c) 
csquareroot = cmath.sqrt(b*b-4*a*c) # Complex number

if squareroot > 0.0: #
    
    print("Gives two real and distinct roots")
    r1=(-b-squareroot)/(2*a) 
    r2=(-b+squareroot)/(2*a)
    print(r1,r2)
    
elif squareroot == 0.0: #
    print("Gives one real root (or two same roots)")
    r1=(-b-squareroot)/(2*a) 
    r2=(-b+squareroot)/(2*a)
    print(r1,r2)
    
else:
    r1=(-b-csquareroot)/(2*a) 
    r2=(-b+csquareroot)/(2*a)
    print("Gives two Complex Roots")
    #print(- b / (2 * a), + i, squareroot)
    #print(- b / (2 * a), - i, squareroot)
    print(r1,r2)





Enter a: 1
Enter b: 2
Enter c: 3
Gives two Complex Roots
(-1-1.4142135623730951j) (-1+1.4142135623730951j)


/Users/ramlijufar/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:5: RuntimeWarning: invalid value encountered in sqrt
  """


## Notice the above warning that means there is some issue which needs fixing, and that tells the above code must be improved


### Let us rather build a Solver where we input the values simultaneously, for accuracy we will add tolerance in this case `tol`. We want the function to display the equation (we can do this using format) along with the solution for different cases of the discriminant.


To use the next function run it and call it with arguments a,b,c which are numbers of your choice to test the different cases. The way to call it in the next cell write `quadratic_solution(1,7,4, tol = 1e-8)`ansd execute the cell. Actuall since the tolerance is fixed you can simply call your function as `quadratic_solution(1,7,4,)`

In [57]:


def quadratic_solution(a,b,c, tol = 1e-7):
    print('Your Equation is: ({0})x^2 + ({1})x +{2}'.format(a,b,c))
    
    # Restrictive conditions on a, b, and c
    if a==b==0:
        if c!=0:
            print('This is not really a valid equation')
        else:
            print('0 = 0 nothing interesting in this equation')
        return
    if a == 0:
        print('Gives one single root r1 =',-c/b)
        return
    
    # checking your discriminant
    inside_squareroot = b**2 - 4*a*c
    
    if inside_squareroot > 0:
        r1 = (-b+math.sqrt(inside_squareroot))/(2*a)
        r2 = (-b-math.sqrt(inside_squareroot))/(2*a)
        print('Gives two real and distinct roots')
        print(r1,r2)
        
    elif inside_squareroot == 0:
        r1 = float(-b+math.sqrt(inside_squareroot))/(2*a)
        r2 = float(-b-math.sqrt(inside_squareroot))/(2*a)
        print('Gives one real root (or two same roots): ')
        print(r1,r2)
        
    
    else:
        r1 = (-b+cmath.sqrt(inside_squareroot))/(2*a)
        r2 = (-b-cmath.sqrt(inside_squareroot))/(2*a)
        print('Gives two Complex Roots')
        print(r1,r2)

In [58]:
quadratic_solution(0,0,0)

Your Equation is: (0)x^2 + (0)x +0
0 = 0 nothing interesting in this equation


In [59]:
quadratic_solution(0,0,2)

Your Equation is: (0)x^2 + (0)x +2
This is not really a valid equation


In [60]:
quadratic_solution(2,6,3)

Your Equation is: (2)x^2 + (6)x +3
Gives two real and distinct roots
-0.6339745962155614 -2.3660254037844384


In [61]:
quadratic_solution(2,4,2)

Your Equation is: (2)x^2 + (4)x +2
Gives one real root (or two same roots): 
-1.0 -1.0


In [62]:
quadratic_solution(1,2,3)

Your Equation is: (1)x^2 + (2)x +3
Gives two Complex Roots
(-1+1.4142135623730951j) (-1-1.4142135623730951j)
